File: Moon/Client/app/service/moon-info.ts
import { BaseResponse } from './base-response';

export enum MoonObjectType {
  Mob,
  Chest,
  Npc
}

export enum MoonMobType {
  Player,
  Mob,
  Npc
}

export enum MoonMobState {
  Await = 0,
  Alive = 1,
  Dead = 2
}

export enum ClientInfoType {
  MobInfo,
  PlayerInfo,
  NpcInfo
}

export enum MoonSkillType {
  // General
  None,
  // Mage
  MageAttack,
  MageFireBall,
  MageFreeze,
  // Knight
  KnightAttack,
  KnightSwordStrike,
  KnightDash,
  // Archer
  ArcherAttack,
  ArcherTripleShot,
  ArcherSlowShot,
  // Priest
  PriestAttack,
  PriestHeal,
  PriestRoot,
  // Mobs
  SkelBowAttack,
  WeakSkelBowAttack,
  TerribleSkelBowAttack,
  SkelAttack,
  SwordSkelAttack,
  PinkZombieAttack
}

export enum SkillState {
  Await,
  Charge,
  Run
}

export enum PlayerClassType {
  Mage,
  Knight,
  Archer,
  Priest
}


export class MoonClientInfoResult extends BaseResponse {
  public infos: Array<ClientInfo>;
}

export enum DebuffType {
  Slow
}

export class MoonDebuff {
  public debuffType: DebuffType;
  public debuffText: string;
  public level: number;
  public speed: number;
  public debuffTime: number;
  public debuffTimeLeft: number;
}

export class MoonDebuffTable {
  public items: Array<MoonDebuff>;
}

export class ClientSprite {
  public spriteName: string;
  public spriteOX: number;
  public spriteOY: number;
  public idleAniSpeed: number;
  public walkAniSpeed: number;
  public runAniSpeed: number;
  public skillAniSpeed: number;
  public deathAniSpeed: number;
}

export class ClientInfo {
  public infoType: ClientInfoType;
  public state: MoonMobState;
  public level: number;
  public awaitTime: number;
  public name: string;
  public mobId: number;
  public targetId: number;
  public targetType: MoonObjectType;
  public sprite: ClientSprite;
  public respMobX: number;
  public respMobY: number;
  public pointMobX: number;
  public pointMobY: number;
  public deltaMobX: number;
  public deltaMobY: number;
  public storeMobX: number;
  public storeMobY: number;
  public runSpeed: number;
  public walkSpeed: number;
  public speed: number;
  public radius: number;
  public maxHP: number;
  public hp: number;
  public pInstruction: string;
  public pInstructionTime: number;
  public pInstructionAllTime: number;
  public pParam1: string;
  public pParam2: number;
  public pParam3: number;
  public directionView: number;
  public skills: Array<SkillInfo>;
  public skillState: SkillState;
  public currentSkillType: MoonSkillType;
  public currentSkillIndex: number
  public chargeTime: number;
  public chargeLeft: number;
  public animateName: string;
  public debuffs: MoonDebuffTable;
  public party: PartyInfo;
}

export class PlayerInfo extends ClientInfo {
  public classType: PlayerClassType;
  public exp: number;
  public maxExp: number;
  public playerId: string;
  public active: boolean;
}

export class SkillInfo {
  public skillType: MoonSkillType;
  public minDistance: number;
  public maxDistance: number;
  public cooldownTime: number;
  public cooldownTimeLeft: number;
  public chargeTime: number;
  public animateTime: number;
  public power: number;
  public autoAttack: boolean;
}

export class PartyInfo {
  public leader: string;
  public items: Array<string>;
}

export class MoonCharListResult extends BaseResponse {
  public chars: Array<MoonChar>;
}
export class MoonCharResult extends BaseResponse {
    public char: MoonChar;
}
export class MoonCharGuidResult extends BaseResponse {
    public charGuid: string;
}
export class MoonChar {
  public moonCharacterGuid: string;
  public userGuid: string;
  public name: string;
  public class: number;
  public level: number;
  public data: string;
  public accessDate: Date;
}

export class LogOnData {
  public login: string;
  public password: string;
}

File: Moon/Server/MoonInfo.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;

namespace Moon
{
  public enum MoonMobState
  {
    Await = 0,
    Alive = 1,    
    Dead = 2 
  }

  public enum ChatType  
  {
    TabAll, 
    TabCommon,
    TabFight,
    TabGroup
  }
  public enum ClientInfoType
  {
    MobInfo,
    PlayerInfo,
    NpcInfo
  }
  public enum MessageBoxButton
  {
    mbOk,
    mbCancel
  }

  public enum MoonObjectType
  {
    Mob,
    Chest,
    Npc
  }

  public class MoonPoint
  {
    public double X, Y;
    public MoonPoint(double X, double Y)
    {
      this.X = X;
      this.Y = Y;
    }
  }

  public class ClientSprite
  {
    public string SpriteName;
    public double SpriteOX = 0;
    public double SpriteOY = 0;
    public double IdleAniSpeed = 3.5;
    public double WalkAniSpeed = 1.6;
    public double RunAniSpeed = 0.8;
    public double SkillAniSpeed = 0.8;
    public double DeathAniSpeed = 1.5;
  }
  public class ClientInfo
  {
    public ClientInfoType InfoType;
    public MoonMobState State;
    public ClientSprite Sprite;
    public double AwaitTime;
    public string Name;
    public int MobId;
    public string ItemCode;
    public int TargetId;
    public MoonObjectType TargetType;
    public double RespMobX;
    public double RespMobY;
    public double PointMobX;
    public double PointMobY;
    public double DeltaMobX;
    public double DeltaMobY;
    public double StoreMobX;
    public double StoreMobY;
    public double RunSpeed;
    public double WalkSpeed;
    public double Speed;
    public double Radius;
    public double MaxHP;
    public double HP;
    public string PInstruction;
    public double PInstructionTime;
    public double PInstructionAllTime;
    public string PParam1;
    public double PParam2;
    public double PParam3;
    public int DirectionView;
    public List<MoonSkill> Skills;
    public SkillState SkillState;
    public int Level;
    public double ChargeTime;
    public double ChargeLeft;
    public MoonSkillType CurrentSkillType;
    public int CurrentSkillIndex;
    public string AnimateName;
    public MoonDebuffTable Debuffs;
    public PartyInfo Party;
    public List<QuestNpcInfo> QuestNpcInfo;

    public ClientInfo()
    {
      InfoType = ClientInfoType.MobInfo;
      Skills = new List<MoonSkill>();
      Sprite = new ClientSprite();
    }
  }

  public class PlayerInfo : ClientInfo
  {
    public Guid PlayerId;
    public bool Active;
    public PlayerClassType ClassType;
    public double Exp;
    public double MaxExp;
    public PlayerInfo()
    {
      InfoType = ClientInfoType.PlayerInfo;
    }
  }

  public class NpcInfo : ClientInfo
  {
    public NpcInfo()
    {
      InfoType = ClientInfoType.NpcInfo;
    }
  }

  public class SkillInfo
  {
    public MoonSkillType SkillType;
    public int MinDistance;
    public int MaxDistance;
    public double CooldownTime;
    public double ChargeTime;
    public double AnimateTime;
    public double Power;
  }

  public class PartyInfo
  {
    public Guid Leader;
    public List<Guid> Items;
  }


}
